File: components/lib/recipe.ts
import { Hit } from "@/src-tauri/bindings/Hit";
import { RecipeData } from "@/src-tauri/bindings/RecipeData";
import { RecipeDetails } from "@/src-tauri/bindings/RecipeDetails";

export function processRecipeData(edamamHit: Hit, recipeScrapersData: RecipeData): RecipeDetails {
  return {
    image: edamamHit.recipe.image,
    url: edamamHit.recipe.url,
    servings: convertToInteger(edamamHit.recipe.yield),
    time: convertToInteger(edamamHit.recipe.totalTime),
    calories: convertToInteger(edamamHit.recipe.calories),
    source: recipeScrapersData.host,
    ingredients: recipeScrapersData.ingredients,
    instructions: recipeScrapersData.instructions_list,
    title: edamamHit.recipe.label,
    id: null,
    date_added: null,
  };
}

function convertToInteger(floatValue: number) {
  return Math.round(floatValue);
}

File: src-tauri/src/types/recipe.rs
use serde::{Deserialize, Serialize};
use ts_rs::TS;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[serde(rename_all = "camelCase")]
#[ts(export)]
pub struct RecipeData {
    #[serde(rename = "canonical_url")]
    pub canonical_url: Option<String>,
    pub category: Option<String>,
    pub host: Option<String>,
    pub image: Option<String>,
    pub ingredients: Option<Vec<String>>,
    pub instructions: Option<String>,
    #[serde(rename = "instructions_list")]
    pub instructions_list: Option<Vec<String>>,
    pub language: Option<String>,
    pub ratings: Option<f64>,
    #[serde(rename = "site_name")]
    pub site_name: Option<String>,
    pub title: Option<String>,
    pub yields: Option<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize, TS)]
#[ts(export)]
pub struct RecipeDetails {
    pub image: String,
    pub url: String,
    pub servings: i32,
    pub time: i32,
    pub calories: i32,
    pub source: Option<String>,
    pub ingredients: Option<Vec<String>>,
    pub instructions: Option<Vec<String>>,
    pub title: String,
    pub id: Option<i32>,
    pub date_added: Option<String>,
}
